aeppl/truncation.py

```diff
@@ -123,3 +131,127 @@ def censor_logprob(op, values, base_rv, lower_bound, upper_bound, **kwargs):
         )
 
     return logprob
+
+
+@singledispatch
+def _truncated(op: Op, lower, upper, *params):
+    """Return the truncated equivalent of antoher ``RandomVariable``.
+
+    The returned Variable should also be a ``RandomVariable``
+    """
+    raise NotImplementedError(
+        f"{op} does not have an equivalent truncated version implemented"
+    )
+
+
+@_truncated.register(arb.UniformRV)
+def uniform_truncated(op, lower, upper, rng, size, dtype, lower_orig, upper_orig):
+    return at.random.uniform(
+        at.max((lower_orig, lower)),
+        at.min((upper_orig, upper)),
+        rng=rng,
+        size=size,
+        dtype=dtype,
+    )
+
+
+class TruncatedRV(OpFromGraph):
+    """An `Op` constructed from an Aesara graph that represents a truncated univariate RV."""
+
+    default_output = 1
+
+
+def truncate(rv, lower=None, upper=None, max_n_steps=10_000):
+
+    if not rv.owner and isinstance(rv.owner.op, RandomVariable):
+        raise ValueError("truncate rv must be a RandomVariable")
+    if rv.owner.op.ndim_supp > 0:
+        raise NotImplementedError(
+            "truncation not implemented for multivariate variables"
+        )
+    if lower is None and upper is None:
+        raise ValueError("lower and upper cannot both be None")
+
+    lower = at.as_tensor_variable(lower) if lower is not None else at.constant(-np.inf)
+    upper = at.as_tensor_variable(upper) if upper is not None else at.constant(np.inf)
+    graph_inputs = [*rv.owner.inputs, lower, upper]
+
+    # Variables with `_` suffix identify dummy inputs for the OpFromGraph
+    graph_inputs_ = [inp.type() for inp in graph_inputs]
+    *rv_inputs_, lower_, upper_ = graph_inputs_
+
+    try:
+        # Use specialized Op
+        truncated_rv_ = _truncated(rv.owner.op, lower_, upper_, *rv_inputs_)
+        return TruncatedRV(
+            inputs=graph_inputs_,
+            outputs=truncated_rv_.owner.outputs,
+            inline=True,
+        )(*graph_inputs)
```

Comment:
Probably it doesn't make sense to wrap these in a `TruncatedRV`, they are supposed to be fully fledged `Ops` with their own `logp` implementations. 

Here, I intended to use the `truncexpon` which is the only truncated implementation in `Aesara` but that one only allows for truncation from above. The example I added here for test with the `uniform` is arguably silly. Other cases would be PyMC's `truncated_normal` which wraps around Scipy's implementation.

Reply:
We might need some sort of dispatch for `Op` truncation cases with existing/efficient implementations.

Reply:
I actually did that in my latest push. It short circuits if there's a dispatched implementation